server/controllers/finance/reports/balance_sheet/index.js

```diff
@@ -0,0 +1,256 @@
+ * @description process and format accounts balance
+ * @param {object} balances The result of balanceReporting function
+ */
+function processAccounts(data) {
+  const collection = Object.keys(data);
+  const bundle = {};
+
+  collection.forEach((type) => {
+    const balances = data[type];
+
+    const accounts = balances.reduce((account, row) => {
+      const id = row.number;
+      const obj = {};
+      account[id] = obj;
+      const sold = getSold(row);
+      obj.label = row.label;
+      obj.number = row.number;
+      obj.debit = sold.debit;
+      obj.credit = sold.credit;
+      obj.balance = sold.debit - sold.credit;
+      obj.is_charge = row.is_charge;
+      obj.is_asset = row.is_asset;
+      return account;
+    }, {});
+
+    // process for getting totals
+    const totals = Object.keys(accounts)
+    .reduce((t, key) => {
+      const account = accounts[key];
+      t.debit += (account.debit || 0);
+      t.credit += (account.credit || 0);
+      t.balance += (account.balance || 0);
+      return t;
+    }, {
+      debit   : 0,
+      credit  : 0,
+      balance  : 0,
+    });
+
+    bundle[type] = { accounts, totals };
+  });
+
+  return bundle;
+}
+
+/**
+ * GroupAccountByType
+ * @description group accounts by account type
+ * @param {array} rows - data from computeBalanceSheet
+ * @returns {object}
+ */
+function GroupAccountByType(rows) {
+  return _.groupBy(rows, 'type_id');
+}
+
+/**
+ * @function getSold
+ * @description return the balance of an account
+ * @param {object} object An object from the array returned by computeBalanceSheet function
+ */
+function getSold(item) {
+  let debit = 0;
+  let credit = 0;
+  let sold = 0;
+
+  if (item.type_id === ASSET || item.type_id === LIABILITY) {
+    sold = item.debit - item.credit;
+    if (sold < 0) {
+      credit = sold * -1;
+    } else {
+      debit = sold;
+    }
+  } else {
+    sold = item.credit - item.debit;
+    if (sold < 0) {
+      debit = sold * -1;
+    } else {
+      credit = sold;
+    }
+  }
+  return { debit, credit };
+}
+
+/**
+ * @function computeBalanceSheet
+ * @description return the balance sheet data according given parameters
+ * @param {object} params An object which contains dates range and the account class
+ */
+function computeBalanceSheet(params) {
+  const query = params;
+
+  query.date = moment(query.date).format(DATE_FORMAT);
+
+  // gets the amount up to the current period
+  const sql = `
+    SELECT a.number, a.id, a.label, a.type_id,
+      SUM(pt.credit) AS credit, SUM(pt.debit) AS debit, SUM(pt.debit - pt.credit) AS balance 
+    FROM period_total AS pt JOIN account AS a ON pt.account_id = a.id
+    JOIN period AS p ON pt.period_id = p.id
+    WHERE pt.enterprise_id = ?  
+      AND (DATE(p.start_date) <= DATE(?) OR (p.start_date IS NULL OR p.end_date IS NULL))
+      AND pt.fiscal_year_id = (SELECT f.id FROM fiscal_year f WHERE DATE(?) BETWEEN DATE(f.start_date) AND DATE(f.end_date) LIMIT 1)
+    GROUP BY a.id `;
```

Comment:
It would be best to do a lot of the sum/ totals in MySQL if it is possible. 

I suggest adding a variable to change the grouping. 

For example 
```sql
GROUP BY account_id #returns account balances
```
```sql 
GROUP BY type_id #returns category balances 
```
```sql 
GROUP BY NULL #returns flat balance (all accounts) 
```

You can also use ```ABS()``` to return a positive balance.